broker/broker.go

```diff
@@ -13,32 +13,129 @@
 package broker
 
 import (
+	"encoding/json"
 	"errors"
 	"fmt"
 	"io/ioutil"
 	"log"
 	"os"
 	"path/filepath"
+	"sync"
+	"time"
 
+	"github.com/eclipse/che-plugin-broker/cfg"
 	"github.com/eclipse/che-plugin-broker/model"
 	"github.com/eclipse/che-plugin-broker/storage"
+	"github.com/eclipse/che/agents/go-agents/core/event"
+	"github.com/eclipse/che/agents/go-agents/core/jsonrpc"
 	yaml "gopkg.in/yaml.v2"
 )
 
+var (
+	bus = event.NewBus()
+)
+
 func ProcessPlugins(metas []model.PluginMeta) {
 	for _, meta := range metas {
 		err := processPlugin(meta)
 		if err != nil {
-			log.Panic(err)
+			log.Fatal(err)
 		}
 	}
 
 	log.Println("Set success status")
 	if ok, status := storage.SetStatus(model.StatusDone); !ok {
-		log.Panicf("Setting '%s' broker status failed. Broker has '%s' state", model.StatusDone, status)
+		log.Fatalf("Setting '%s' broker status failed. Broker has '%s' state", model.StatusDone, status)
+	}
+}
+
+func Start(metas []model.PluginMeta) {
+	if ok, status := storage.SetStatus(model.StatusStarting); !ok {
+		m := fmt.Sprintf("Starting broker in state '%s' is not allowed", status)
+		pubFailed(m)
+		log.Fatal(m)
+	}
+
+	for _, meta := range metas {
+		err := processPlugin(meta)
+		if err != nil {
+			pubFailed(err.Error())
+			log.Fatal(err)
+		}
+	}
+
+	log.Println("Set success status")
+	if ok, status := storage.SetStatus(model.StatusDone); !ok {
+		err := fmt.Sprintf("Setting '%s' broker status failed. Broker has '%s' state", model.StatusDone, status)
+		pubFailed(err)
+		log.Fatalf(err)
+	}
+
+	tooling, err := storage.Tooling()
+	if err != nil {
+		pubFailed(err.Error())
+		log.Fatalf(err.Error())
+	}
+	bytes, err := json.Marshal(tooling)
+	if err != nil {
+		pubFailed(err.Error())
+		log.Fatalf(err.Error())
+	}
+	pubDone(string(bytes))
+	closeConsumers()
+}
+
+func closeConsumers() {
+	for _, candidates := range bus.Clear() {
+		for _, candidate := range candidates {
+			if broadcaster, ok := candidate.(*tunnelBroadcaster); ok {
+				broadcaster.Close()
+			}
+		}
 	}
 }
 
+func (tb *tunnelBroadcaster) Close() { tb.tunnel.Close() }
+
+func pubFailed(err string) {
+	bus.Pub(&model.ErrorEvent{
+		Status:      model.StatusFailed,
+		Error:       err,
+		WorkspaceID: cfg.WorkspaceID,
+	})
+}
+
+func pubDone(tooling string) {
+	bus.Pub(&model.SuccessEvent{
+		Status:      model.StatusDone,
+		WorkspaceID: cfg.WorkspaceID,
+		Tooling:     tooling,
+	})
+}
+
+// PushStatuses sets given tunnel as consumer of broker events.
+func PushStatuses(tun *jsonrpc.Tunnel) {
+	bus.SubAny(&tunnelBroadcaster{tunnel: tun}, model.BrokerStatusEventType, model.BrokerResultEventType)
+}
+
+type tunnelBroadcaster struct {
+	tunnel          *jsonrpc.Tunnel
```

Comment:
everything except tunnel is unused in this struct. Is this intentionally?

Reply:
It was copied by mistake. Reconnection feature is not implemented in the broker. I've removed this code.